sway/commands/move.c

```diff
@@ -263,138 +265,141 @@ static void container_move_to_container(struct sway_container *container,
-	child->width = child->height = 0;
+	return false;
 }
 
 // Returns true if moved
 static bool container_move_in_direction(struct sway_container *container,
 		enum wlr_direction move_dir) {
 	// If moving a fullscreen view, only consider outputs
-	if (container->fullscreen_mode == FULLSCREEN_WORKSPACE) {
-		struct sway_output *new_output =
-			output_get_in_direction(container->workspace->output, move_dir);
-		if (!new_output) {
-			return false;
-		}
-		struct sway_workspace *ws = output_get_active_workspace(new_output);
-		if (!sway_assert(ws, "Expected output to have a workspace")) {
-			return false;
-		}
-		container_move_to_workspace(container, ws);
-		return true;
-	}
-	if (container->fullscreen_mode == FULLSCREEN_GLOBAL) {
+	switch (container->fullscreen_mode) {
+	case FULLSCREEN_NONE:
+		break;
+	case FULLSCREEN_WORKSPACE:
+		return container_move_to_next_output(container,
+				container->workspace->output, move_dir);
+	case FULLSCREEN_GLOBAL:
 		return false;
 	}
 
-	// If container is in a split container by itself, move out of the split
-	if (container->parent) {
-		struct sway_container *old_parent = container->parent;
-		struct sway_container *new_parent =
-			container_flatten(container->parent);
-		if (new_parent != old_parent) {
-			return true;
-		}
-	}
-
-	// Look for a suitable *container* sibling or parent.
-	// The below loop stops once we hit the workspace because current->parent
-	// is NULL for the topmost containers in a workspace.
-	struct sway_container *current = container;
 	int offs =
 		move_dir == WLR_DIRECTION_LEFT || move_dir == WLR_DIRECTION_UP ? -1 : 1;
+	int index = -1;
+	int	desired = -1;
+	list_t *siblings = NULL;
+	struct sway_container *target = NULL;
+
+	// Look for a suitable ancestor of the container to move within
+	struct sway_container *ancestor = NULL;
+	struct sway_container *current = container;
+	bool wrapped = false;
+	while (!ancestor) {
+		// Don't allow containers to move out of their
+		// fullscreen or floating parent
+		if (current->fullscreen_mode || container_is_floating(current)) {
+			return false;
+		}
 
-	while (current) {
-		list_t *siblings = container_get_siblings(current);
-		if (siblings) {
-			enum sway_container_layout layout = container_parent_layout(current);
-			int index = list_find(siblings, current);
-			int desired = index + offs;
-
-			// Don't allow containers to move out of their
-			// fullscreen or floating parent
-			if (current->fullscreen_mode || container_is_floating(current)) {
-				return false;
+		enum sway_container_layout parent_layout = container_parent_layout(current);
+		if (!is_parallel(parent_layout, move_dir)) {
+			if (!current->parent) {
+				// No parallel parent, so we reorient the workspace
+				current = workspace_wrap_children(current->workspace);
+				current->workspace->layout =
+					move_dir == WLR_DIRECTION_LEFT ||
+					move_dir == WLR_DIRECTION_RIGHT ?
+					L_HORIZ : L_VERT;
+				container->height = container->width = 0;
```

Comment:
Somewhat an aside from the meat of this PR, but I think we should have something like a `container_reset_geometry` for this since these lines appear in a lot of other places.

Reply:
You're probably right, but I think I'll hold off adding that in this PR.